include/nektar_interface/particle_cell_mapping/generated_linear/hexahedron.hpp

```diff
@@ -0,0 +1,277 @@
+ *
+ * @param[in] xi0 Current xi_n point, x component.
+ * @param[in] xi1 Current xi_n point, y component.
+ * @param[in] xi2 Current xi_n point, z component.
+ * @param[in] v00 Vertex 0, x component.
+ * @param[in] v01 Vertex 0, y component.
+ * @param[in] v02 Vertex 0, z component.
+ * @param[in] v10 Vertex 1, x component.
+ * @param[in] v11 Vertex 1, y component.
+ * @param[in] v12 Vertex 1, z component.
+ * @param[in] v20 Vertex 2, x component.
+ * @param[in] v21 Vertex 2, y component.
+ * @param[in] v22 Vertex 2, z component.
+ * @param[in] v30 Vertex 3, x component.
+ * @param[in] v31 Vertex 3, y component.
+ * @param[in] v32 Vertex 3, z component.
+ * @param[in] v40 Vertex 4, x component.
+ * @param[in] v41 Vertex 4, y component.
+ * @param[in] v42 Vertex 4, z component.
+ * @param[in] v50 Vertex 5, x component.
+ * @param[in] v51 Vertex 5, y component.
+ * @param[in] v52 Vertex 5, z component.
+ * @param[in] v60 Vertex 6, x component.
+ * @param[in] v61 Vertex 6, y component.
+ * @param[in] v62 Vertex 6, z component.
+ * @param[in] v70 Vertex 7, x component.
+ * @param[in] v71 Vertex 7, y component.
+ * @param[in] v72 Vertex 7, z component.
+ * @param[in] phys0 Target point in global space, x component.
+ * @param[in] phys1 Target point in global space, y component.
+ * @param[in] phys2 Target point in global space, z component.
+ * @param[in] f0 Current f evaluation at xi, x component.
+ * @param[in] f1 Current f evaluation at xi, y component.
+ * @param[in] f2 Current f evaluation at xi, z component.
+ * @param[in, out] xin0 Output local coordinate iteration, x component.
+ * @param[in, out] xin1 Output local coordinate iteration, y component.
+ * @param[in, out] xin2 Output local coordinate iteration, z component.
+ */
+inline void newton_step_linear_3d(
+    const REAL xi0, const REAL xi1, const REAL xi2, const REAL v00,
+    const REAL v01, const REAL v02, const REAL v10, const REAL v11,
+    const REAL v12, const REAL v20, const REAL v21, const REAL v22,
+    const REAL v30, const REAL v31, const REAL v32, const REAL v40,
+    const REAL v41, const REAL v42, const REAL v50, const REAL v51,
+    const REAL v52, const REAL v60, const REAL v61, const REAL v62,
+    const REAL v70, const REAL v71, const REAL v72, const REAL phys0,
+    const REAL phys1, const REAL phys2, const REAL f0, const REAL f1,
+    const REAL f2, REAL *xin0, REAL *xin1, REAL *xin2) {
+  const REAL x0 = xi1 - 1;
+  const REAL x1 = xi2 - 1;
+  const REAL x2 = x0 * x1;
+  const REAL x3 = xi1 + 1;
+  const REAL x4 = x1 * x3;
+  const REAL x5 = xi2 + 1;
+  const REAL x6 = x0 * x5;
+  const REAL x7 = x3 * x5;
+  const REAL x8 = xi0 - 1;
+  const REAL x9 = x1 * x8;
+  const REAL x10 = xi0 + 1;
+  const REAL x11 = x1 * x10;
+  const REAL x12 = x10 * x5;
+  const REAL x13 = x5 * x8;
+  const REAL x14 = x0 * x8;
+  const REAL x15 = x10 * x3;
+  const REAL x16 = x0 * x10;
+  const REAL x17 = x3 * x8;
+  const REAL J00 = -0.125 * v00 * x2 + 0.125 * v10 * x0 * x1 -
```

Comment:
It looks as though these expressions have passed through a common subexpression elimination tool? If so let's make 0.125 a variable (one_8 or whatever) so that we can eliminate lots of multiplies.

Reply:
I don't think this gains anything for the additional work. 

On the readability side: this isn't C you would particularly read without being very familiar with what is happening already so "one_8" vs 0.125 is the same. On the optimisation side explicit inlining of constants, is generally better than using a variable (especially if these ops were integers where the compiler will emit specific instructions depending on the value).  

The compiler will do a second CSE pass and (hopefully) rearrange the whole thing anyway depending on whether it decides (0.125 * x) * y is better than 0.125 * (x * y). This CSE should be thought of as a "first pass".